### DISCUSSION TITLE
Project settings' Custom JavaScript

### DISCUSSION THREAD

**Discussion by Pachat** [directus/directus#12266] [Feature Requests]
In the manner we have Project settings' Custom CSS, would it be possible to have Project settings' Custom JavaScript?

This would allow to set events in the App without having to rebuild 99% of what Directus already does very well just to add some logic.

_Or is there already a way that I missed?_

**Comment by rijkvanzanten** [comment]
> Or is there already a way that I missed?

Besides the custom extensions, there's no way to run any (global) javascript through an overrides field like the css overrides at the moment

**Comment by ched-dev** [comment]
This would really open a can of worms for security issues. What do you mean by:
> This would allow to set events in the App

Or in other words, what is the exact feature you are trying to add in a theoretical Custom JS field?

**Reply by Pachat** [reply]
The exact feature to add is an event on closing a section in an accordion in order to send programmatically `saveAndStay()` on long forms _that does not have any required field and therefore can be "Saved and Stay" at any time_.

Motivation:

On long forms (say 10 groups of 8 inputs each) the user may/**will** forget to do intermediate savings because

- he would have to click from time to time to open the upper menu and click again on the Save and Stay
- he does not know or cannot do `Ctrl+S` (tablet?) or preempted shortcut by the navigator.
- he may does not know the feature or even what it means (yes!)

_In a word, he is concentrated on feeding the form and should not have to bother about intermediate savings._

Accordion is a nice feature:

- Beginning with all sections closed shows the list of the sections of the form
- which allows easily to go directly to the specific part where he wants to continue feeding the form
- using `Max 1 Section Open` will close the current section and is a good event to throw programmatically a `saveAndStay()`

**Reply by Pachat** [reply]
Another use case is that I want some Dropdonws to be shown opened without having to click on the `v`.
This is possible with Vue, but require Javascript:

[dropdownShouldOpen v3](https://vue-select.org/api/props.html#dropdownshouldopen)

> Determines whether the dropdown should open. Used for overriding the default dropdown behaviour. Receives the vue-select instance as the single argument to the function.

```
dropdownShouldOpen: {
    type: Function,
    default({noDrop, open, mutableLoading}) {
      return noDrop ? false : open && !mutableLoading;
    }
}
```

**Reply by kamil-malinski** [reply]
I second that, it would be very useful if something like that was possible

**Comment by diegoleme** [comment]
I would like to include third-party scripts to monitor the Directus app in the browser. (Ex.: Sentry, New Relic, ELK...)

**Comment by diegoleme** [comment]
If that makes sense, can I try to open a PR?

**Comment by nickrum** [comment]
The usefulness of this would be very limited because the App is bundled and minified when served in production, which means all of the variable names are mangled. Also, as the App is bundled to ESM, the code of the App is not available globally but isolated inside modules.

**Reply by diegoleme** [reply]
My need is to inject monitoring codes, in my case I use ELK. And custom js code is one way to do that.

**Reply by nickrum** [reply]
Rather than allowing for custom js, I think we should add a `global` App extension type that would allow you to inject code into the components and elements of the app.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by alexey-yarmosh** [comment]
On our project we would like to add small UI changes to some of the pages. Specifically we want to hide search button for the collection 'a' page, but remain the search for all other pages. Currently there is no option to stay with the default collection view and slightly update it. Custom css applies to every page so that is no-op. Custom js should be extra helpful for such things.

**Reply by Pachat** [reply]
To work around the lack of custom.js, we add either an interface or a display extension on a field we added to the collection on that purpose. The field is loaded but does not display anything by itself. 
The `<script>...</script>` applies to the page where it is called. 
Not ideal, but it works.

**Reply by br41nslug** [reply]
Using an `embed hook` you can inject javascript directly into the page https://docs.directus.io/extensions/hooks.html#embed

Beware that making "small" changes to generated code can be tricky (think of JS page changes, generated classes etc)

**Reply by Pachat** [reply]
@br41nslug Sounds the best way :). I suppose we do have access to the api there.

**Reply by alexey-yarmosh** [reply]
Thanks for help, hook seems to be the perfect option. Btw we were able to solve our issue using complex css selector ‘body:not(:has(.router-link-active[href="/admin/content/a"])) .search-input’

**Reply by br41nslug** [reply]
> @br41nslug Sounds the best way :). I suppose we do have access to the api there.

The embeds run in the browser context (same as if you were to run code directly in the dev panel) but you dont have direct access to Directus internals like the current access_token etc.

**Reply by Pachat** [reply]
Thanks for the precision.

**Comment by abernh** [comment]
For selfhosting this is absolutely not needed as all the functionality can be created via custom extensions.

But on Directus Cloud you are completely limited and even the smallest of adjustments to pages are restricted to what the presets allow, e.g. some customization to sidebars would always be helpful.

And there is no need for the `auth_token` to be available as all JS will be running on the same domain and therefore with the same Cookie auth - so yes, the custom JS will be fully able to access the API and bring a huge amount of functionality with it that otherwise would not be possible (in Directus Cloud (Non-Enterprise)).

Also there are still too many little bugs (like the missing custom error message on new-password) that could simply be fixed by a custom script.
Open source only is helpful if I am also allowed to participate, even if it is only my own instance - and Directus Cloud is currently restricting any participation to custom CSS only and that is simply a Devs nightmare come true.

**Reply by Pachat** [reply]
As there is "no free lunch" even in Open Source, just know that as far as I can tell the Directus team that does a great job of opening it up to you is also very open to contributions. Besides that, these might be good reasons why, so far, they haven't implemented custom JavaScript.

**Reply by abernh** [reply]
Yes, the Directus team does a great job on allowing participation as I have witnessed already some of my own PRs for minor bug fixes getting a merge. 
Maybe I'm just impatient / frustrated with sitting in the "golden cage" of a no-sys-op-necessary cloud instance which in return I am just a tad too limited for adjustments and imagining this simple settings field with custom JS being the fix for all my issues.

**Reply by rijkvanzanten** [reply]
I have my doubt that the custom JS field would actually allow you to achieve the type of stuff you're thinking of. Keep in mind that (like custom css) it would execute once on page load, and it's up to you to handle things from there. It wouldn't be able to tap into the Vue lifecycle or page transitions or anything like that. 

I don't think custom JS would be able to be used for little bugs like the example you mentioned:

> [...] like the missing custom error message on new-password) that could simply be fixed by a custom script.

**Reply by abernh** [reply]
> It wouldn't be able to tap into the Vue lifecycle or page transitions or anything like that.

Yes, adjusting the layout (without access to the Vue lifecycle and especially via CSS only) is a real pain in Directus because thanks to the heavily used `style[scoped]` approach in all the components there are hardly any "telling" class names around to work with. Same problem with the `props`-only approach where every `input` field looks the same because they don't even have a `[name]` attribute anymore.

But there is no need to use the Vue livecycle once you are with your back to a wall and the customer needs a change.
Checking "on-page-load/location-change" for "where-we-are" gives already already a good base and then it is only a matter of selector finesse  and some well placed (! to avoid a performance nightmare this needs to be as surgical as possible) MutationObservers to keep you "in the loop" of the latest state and you are again king of the layout. 
So changing the message of an error box or simply adding a tooltip / instruction block (like for the password reset form) would be much easier with JS support (compared to [brittle CSS only solutions](https://github.com/directus/directus/discussions/13896#discussioncomment-6275872)).

Directus is a really awesome tool and it allows you to create impressive interfaces even without all the extensions. 

But customer satisfaction really reaches a new high when I can also deliver the little details they are listing and generally that is a no brainer given all the easy extension functionalities. But for the singular situation of the Directus Cloud (NonEnterprise) it would just be nice to have this custom JS field to get a "foot in the door" of the layout.

I also tried to setup only the Studio App selfhosted and connect it to the cloud instance. That worked flawless, but it forces me to keep my selfhosted frontend in sync with the automatically updating cloud instance; something I'd like to avoid if it is only because of some minor gains.
